### PR TITLE
Make sure a constructor in extension cannot steal the initializing_with_extension==true status

### PR DESCRIPTION
**Pull request by groud**
While trying to work with godot-cpp (investigating a crash), I figured out that `instance_get_native_extension_data` would not be called with the correct object when `initializing_with_extension` was true.

To find out that, I printed `this` in my object constructor and `r_extension_instance` in `instance_get_native_extension_data`. Before this PR, the pointers differed, while now, they are the same.

The issue was due to the extension object constructor being called in `ti->native_extension->create_instance`. If any object was instantiated in this constructor, it would take over the global variable `initializing_with_extension == true` status, and be initialized as an extension object even if it is not. By moving `initializing_with_extension = true;` afterward, we avoid this problem.

### CODE REVIEW

File: core/object/class_db.cpp [hunk 1]
```diff
@@ -534,9 +534,10 @@ Object *ClassDB::instantiate(const StringName &p_class) {
 	}
 #endif
 	if (ti->native_extension) {
-		initializing_with_extension = true;
-		initializing_extension = ti->native_extension;
 		initializing_extension_instance = ti->native_extension->create_instance(ti->native_extension->class_userdata);
+		initializing_extension = ti->native_extension;
```
Comment (BastiaanOlij) [inline core/object/class_db.cpp:538]: I think the same fix needs to be done in ClassDB::construct_object

**Comment by BastiaanOlij** [comment]
in `ClassDB::instance_get_native_extension_data` I would also clear the pointers, so:
```
void ClassDB::instance_get_native_extension_data(ObjectNativeExtension **r_extension, GDExtensionClassInstancePtr *r_extension_instance, Object *p_base) {
	if (initializing_with_extension) {
		*r_extension = initializing_extension;
		*r_extension_instance = initializing_extension_instance;
		initializing_extension->set_object_instance(*r_extension_instance, p_base);

		initializing_with_extension = false;
		initializing_extension = nullptr;
		initializing_extension_instance = nullptr;
	} else {
		*r_extension = nullptr;
		*r_extension_instance = nullptr;
	}
}
```

**Comment by BastiaanOlij** [comment]
Ok, now the sauce and deep dive of what we're working around here and what the deeper problem is that requires a redesign of what is happening here. Both me and @groud ran into this problem, investigated it independently and came to the same conclusion.

The problem is that `instance_get_native_extension_data` was being called before the object in question was fully constructed and that made no sense.

Then we realised that in both our codes we were doing this in our constructor:
```
UtilityFunctions::print("...")
```

Seems innocent enough and it indeed is, in runtime, but in the editor this doesn't just log a message, it actually displays that message in the editor, which in Godot is done by creating new text elements in the debug log window. that means new objects, that means new constructors, and those constructors were trying to get the global info we're grabbing here.

Not setting the flag until after OUR constructor has finished works around this, but it will ONLY work if our extension objects constructor doesn't end up constructing other extension objects. Thats an unlikely but not impossible scenario.

So this solution will work around the problem and be enough in 99.9% of the cases, but using global variables for this handover is not a safe approach and while I vote for merging these fixes ASAP,we should investigate if we can replace this with a much safer approach.

**Comment by groud** [comment]
To also complete my investigations here is my explanation on why this code is the way it is (copyied from a message in RocketChat):

I think the goal of instance_get_native_extension_data() is get the extension pointers as soon as possible in the constructors chain. As in C++, constructors are called from parent classes to children classes, having it in the Object constructor's makes sure it is initialized as soon as possible. 

And I think there's not a lot of other solutions, if you want to setup an object as the first step of it's constructor, you don't have a lot of solutions either you change the constructor arguments so that it handles the data you pass it, but it means you have to modify all children constructor, or you use a global variable, and let the constructor retrieve the data itself.

The later solution was chosen, as it is kind of simpler code, but the only thing we have to make sure, is that the global variable in question won't be modified at all before it is used. This did happen in our case during the initialization of the extension_object constructor, before this PR.

So with this fix, I think the `initializing_with_extension = true` is the closer we can get to the `instance_get_native_extension_data()` call, as it is ran in the Godot object constructor, the one created by `ti->create_func()`
I don't think anything can cause the variables to be messed up there anymore (but time will tell us).

**Comment by BastiaanOlij** [comment]
I wonder if on the `godot-cpp` side in our `Object::Object()` constructor, or maybe even in `Wrapped`, if we can do an immediate callback into Godot to inform it of the new pointer and do the setup right then and there.

**Comment by groud** [comment]
I'll mark this as WIP, as it still needs some work it seems. It might cause a crash.

**Comment by BastiaanOlij** [comment]
Seeing you're well on your way to writing the replacement for this and I've got a temporary fix in place, just keep it here for reference and trash it when the other work is done?

**Comment by groud** [comment]
Closing as I completely changed the branch content, so that extension initialize the Godot object, and not the other way around.

**Pull request closed without merging**